torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+
+            state = status["state"]
+            app_state = JOB_STATE[state]
+
+            # Roles
+            spec = api_resp["spec"]
+            resources = spec["resources"]
+            genericItems = resources["GenericItems"]
+
+            # Note MCAD service is not considered a TorchX role
+            roles = get_role_information(genericItems)
+            task_count = 0
+            for role in roles:
+                msg = "Warning - MCAD does not report individual replica statuses, but overall task status. Replica id  may not match status"
+                warnings.warn(msg)
+
+                roles_statuses[role] = RoleStatus(role, [])
+                for idx in range(0, roles[role].num_replicas):
+                    state = TASK_STATE[task_status[task_count]]
+                    roles_statuses[role].replicas.append(
+                        ReplicaStatus(id=int(idx), role=role, state=state, hostname="")
+                    )
+                    task_count += 1
+
+        else:
+            app_state = AppState.UNKNOWN
+
+        return DescribeAppResponse(
+            app_id=app_id,
+            roles=list(roles.values()),
+            roles_statuses=list(roles_statuses.values()),
+            state=app_state,
+        )
+
+    def log_iter(
+        self,
+        app_id: str,
+        role_name: str,
+        k: int = 0,
+        regex: Optional[str] = None,
+        since: Optional[datetime] = None,
+        until: Optional[datetime] = None,
+        should_tail: bool = False,
+        streams: Optional[Stream] = None,
+    ) -> Iterable[str]:
+        assert until is None, "kubernetes API doesn't support until"
+
+        if streams not in (None, Stream.COMBINED):
+            raise ValueError(
+                "KubernetesMCADScheduler only supports COMBINED log stream"
+            )
+
+        from kubernetes import client, watch
+
+        namespace, name = app_id.split(":")
+
+        print("\033[100m" + "\n ks - log_iter \n" + "\033[0m")
+        print(f"{name}")
+
+        pod_name = cleanup_str(f"{name}-{k}")
+
+        print("\033[100m" + "\n ks - log_iter \n" + "\033[0m")
+        print(f"{pod_name}")
+
+        args: Dict[str, object] = {
+            "name": pod_name,
+            "namespace": namespace,
+            "timestamps": True,
+        }
+        if since is not None:
+            args["since_seconds"] = (datetime.now() - since).total_seconds()
+
+        core_api = client.CoreV1Api(self._api_client())
+        if should_tail:
+            w = watch.Watch()
+            iterator = w.stream(core_api.read_namespaced_pod_log, **args)
+        else:
+            resp = core_api.read_namespaced_pod_log(**args)
+            iterator = split_lines(resp)
+
+        if regex:
+            return filter_regex(regex, iterator)
+        else:
+            return iterator
+
+    def list(self) -> List[ListAppResponse]:
+        p1 = subprocess.run(
+            ["kubectl", "config", "view", "--minify"],
```

Comment:
this is a clever trick :)

Might be better to grab it from the kube config using the Python API however

https://github.com/kubernetes-client/python/issues/363#issuecomment-389722355

Would be nice to integrate this into the normal k8s scheduler as well

Reply:
Updated MCAD and Volcano schedulers to use Kube config for getting the namespace in both list functions. Commit: bcb0ebe432db59f8ecca3c0ac77bd8bdc2d5a83a